File: integration-tests/testutils/testutils/testutils.go
// -*- Mode: Go; indent-tabs-mode: t -*-
// +build !excludeintegration

/*
 * Copyright (C) 2015 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package testutils

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
)

// PrepareTargetDir creates the given target directory, removing it previously if it didn't exist
func PrepareTargetDir(targetDir string) {
	if _, err := os.Stat(targetDir); err == nil {
		// dir exists, remove it
		os.RemoveAll(targetDir)
	}
	os.MkdirAll(targetDir, 0777)
}

// RootPath return the test current working directory.
func RootPath() string {
	dir, err := os.Getwd()
	if err != nil {
		log.Panic(err)
	}
	return dir
}

// ExecCommand executes the given command and pipes the results to os.Stdout and os.Stderr, returning the resulting error
func ExecCommand(cmds ...string) error {
	fmt.Println(strings.Join(cmds, " "))

	cmd := exec.Command(cmds[0], cmds[1:]...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	err := cmd.Run()
	if err != nil {
		log.Panicf("Error while running %s: %s\n", cmd.Args, err)
	}
	return err
}

// RemoveTestFlags strips the flags beginning with "-test." from os.Args,
// useful for the test binaries compiled as the original cmds for measuring
// integration test coverage
func RemoveTestFlags() {
	outputArgs := []string{}
	for _, item := range os.Args {
		if !strings.HasPrefix(item, "-test.") {
			outputArgs = append(outputArgs, item)
		}
	}
	os.Args = outputArgs
}

File: cmd/snap-update-ns/test-utils.c
/*
 * Copyright (C) 2017 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "test-utils.h"

#include <glib.h>

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

void sc_test_remove_file(const char *name)
{
	int err = remove(name);
	g_assert_cmpint(err, ==, 0);
}

void sc_test_write_lines(const char *name, ...)
{
	FILE *f = fopen(name, "wt");
	g_assert_nonnull(f);

	va_list ap;
	va_start(ap, name);
	const char *line;
	while ((line = va_arg(ap, const char *)) != NULL) {
		fprintf(f, "%s\n", line);
	}
	va_end(ap);
	fclose(f);

	// Cast-away the const qualifier. This just calls unlink and we don't
	// modify the name in any way. This way the signature is compatible with
	// that of GDestroyNotify.
	g_test_queue_destroy((GDestroyNotify) sc_test_remove_file,
			     (char *)name);
}
